drivers/counter/counter_mcux_rtc.c

```diff
@@ -0,0 +1,256 @@
+
+	u32_t ticks = config->base->TSR;
+
+	/*
+	 * Read TSR seconds twice in case it glitches during an update.
+	 * This can happen when a read occurs at the time the register is
+	 * incrementing.
+	 */
+	if (config->base->TSR == ticks) {
+		return ticks;
+	}
+
+	ticks = config->base->TSR;
+
+	return ticks;
+}
+
+static int mcux_rtc_set_alarm(struct device *dev,
+			      const struct counter_alarm_cfg *alarm_cfg)
+{
+	const struct counter_config_info *info = dev->config->config_info;
+	const struct mcux_rtc_config *config =
+		CONTAINER_OF(info, struct mcux_rtc_config, info);
+	struct mcux_rtc_data *data = dev->driver_data;
+
+	u32_t ticks = alarm_cfg->ticks;
+	u32_t current = mcux_rtc_read(dev);
+
+	LOG_DBG("Current time is %d ticks", current);
+
+	if (data->alarm_cfg != NULL) {
+		LOG_ERR("No more alarms available");
+		return -EINVAL;
+	}
+
+	if (!alarm_cfg->absolute) {
+		ticks += current;
+	}
+
+	if (ticks < current) {
+		LOG_ERR("Alarm cannot be earlier than current time");
+		return -EINVAL;
+	}
+
+	data->alarm_cfg = alarm_cfg;
+
+	config->base->TAR = ticks;
+	LOG_DBG("Alarm set to %d ticks", ticks);
+
+	return 0;
+}
+
+static int mcux_rtc_disable_alarm(struct device *dev,
+				  const struct counter_alarm_cfg *alarm_cfg)
+{
+	struct mcux_rtc_data *data = dev->driver_data;
+
+	if (data->alarm_cfg != alarm_cfg) {
+		LOG_ERR("Invalid alarm");
+		return -EINVAL;
+	}
+
+	data->alarm_cfg = NULL;
+
+	return 0;
+}
+
+static int mcux_rtc_set_wrap(struct device *dev, u32_t ticks,
+			     counter_wrap_callback_t callback,
+			     void *user_data)
+{
+	const struct counter_config_info *info = dev->config->config_info;
+
+	if (ticks != info->max_wrap) {
+		LOG_ERR("Wrap can only be set to 0x%x", info->max_wrap);
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static u32_t mcux_rtc_get_pending_int(struct device *dev)
+{
+	const struct counter_config_info *info = dev->config->config_info;
+	const struct mcux_rtc_config *config =
+		CONTAINER_OF(info, struct mcux_rtc_config, info);
+
+	return RTC_GetStatusFlags(config->base) & RTC_SR_TAF_MASK;
+}
+
+static u32_t mcux_rtc_get_wrap(struct device *dev)
+{
+	const struct counter_config_info *info = dev->config->config_info;
+
+	return info->max_wrap;
+}
+
+static u32_t mcux_rtc_get_max_relative_alarm(struct device *dev)
+{
+	const struct counter_config_info *info = dev->config->config_info;
+
+	return info->max_wrap;
+}
+
+static void mcux_rtc_isr(void *arg)
+{
+	struct device *dev = arg;
+	const struct counter_config_info *info = dev->config->config_info;
+	const struct mcux_rtc_config *config =
+		CONTAINER_OF(info, struct mcux_rtc_config, info);
+	struct mcux_rtc_data *data = dev->driver_data;
+	const struct counter_alarm_cfg *alarm_cfg = data->alarm_cfg;
+	u32_t current = mcux_rtc_read(dev);
+
+	LOG_DBG("Current time is %d ticks", current);
+
+	/* perform any registered callbacks */
+	if (alarm_cfg && alarm_cfg->handler) {
+		alarm_cfg->handler(dev, alarm_cfg, current);
```

Comment:
you should return `config->base->TAR` as this should be exact ticks. Current is impacted by interrupt latency.

Reply:
Do you mean the requested ticks or the actual ticks when the callback is invoked? I interpreted the api to mean the latter, because why would you pass a ticks argument when alarm_cfg already contains the requested ticks?